src/extensions/default/Inline3DParametersEditor/main.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * File based on InlineColorEditor/main.js
+ */
+
+define(function (require, exports, module) {
+    "use strict";
+
+    var EditorManager           = brackets.getModule("editor/EditorManager"),
+        ExtensionUtils          = brackets.getModule("utils/ExtensionUtils"),
+        HTMLUtils               = brackets.getModule("language/HTMLUtils"),
+        Inline3dParametersUtils = require("Parameters3DUtils"),
+        Inline3DParameterEditor = require("Inline3DParameterEditor").Inline3DParameterEditor;
+
+    /**
+     * Prepare hostEditor for an Inline3DParameterEditor at pos if possible. Return
+     * editor context if so; otherwise null.
+     *
+     * @param {Editor} hostEditor
+     * @param {{line:Number, ch:Number}} pos
+     * @return {?{parameter:String, marker:TextMarker}}
+     */
+    function prepareParametersForProvider(hostEditor, pos) {
+        var ParameterRegex, tagStart, cursorLine, match, sel, start, end, endPos, marker, tagInfo;
+
+        sel = hostEditor.getSelection();
+        if (sel.start.line !== sel.end.line) {
+            return null;
+        }
+        var tagInfo = HTMLUtils.getTagInfo(hostEditor, sel.start);
+        var tag = tagInfo.tagName;
+        tagStart =Inline3dParametersUtils.TAG_START;
+        if(!tag.substr(0, 2) === tagStart) {
+            return null;
+        }
+
+        ParameterRegex = Inline3dParametersUtils.PARAMETERS_3D_REGEX;
+        ParameterRegex.lastIndex = 0;
+        tagInfo = HTMLUtils.getTagInfo(hostEditor, pos),
+        cursorLine = hostEditor.document.getLine(pos.line);
+        // Loop through each match of ParameterRegex and stop when the one that contains pos is found.
+        do {
+            match = ParameterRegex.exec(cursorLine);
```

Comment:
This was a bit over complicated. I'll try to explain its execution. 
exec function loops through all the instances of match that occur in a string. 
Eg
```js 
cursorLine = "<a-sphere position = "1 2 2" scale = "2 2 2" radius = "3">"
```
And lets say the user placed the cursor at the scale parameter ("2 2 2").
the loop looks something like this
```js
do {
            //exec loops through each match of ParameterRegex in the line cursorLine
            match = ParameterRegex.exec(cursorLine);

            if (match) {
                start = match.index; // start of the match found
                end = start + match[0].length; // end of the match found
            }
        } while (match && (pos.ch < start || pos.ch > end));
```
In the first iteration, 
 match = ParameterRegex.exec(cursorLine); sets match to "1 2 2". However, the end variable in this case is less than the position of the cursor ("2 2 2"). This is what is checked by the while loop. It returns to the next iteration. 
In the next iteration,  match = ParameterRegex.exec(cursorLine); sets match to "2 2 2". In this case start < pos.ch and end >pos.ch. Hence the loop breaks and we get our matched string.

I think the confusion lies in the line 
```js
 match = ParameterRegex.exec(cursorLine); 
```
The value of ParameterRegex.exec doesn't remain same in each iteration. Since ParameterRegex is defined as a global regex, exec loops through all the possible matches as this line is executed. 
Here is a screenshot from the official documentation found [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec).
![screenshot from 2017-04-25 10-11-01](https://cloud.githubusercontent.com/assets/8079861/25369438/83f7571a-29a1-11e7-8b40-8f27729f65b5.png)

However I agree this part required more documentation. Added the same [here](https://github.com/mozilla/brackets/pull/651/commits/e4bd7c078113b76d463f78270e0658e7bc80c258)